File: src/resolvers.ts
import type * as srk from '@algoux/standard-ranklist';
// @ts-ignore
import TEXTColor from 'textcolor';
import { lookup as langLookup } from 'bcp-47-match';
import { EnumTheme } from './enums';
import { ThemeColor } from './types';

/**
 * Resolve an srk text value into a display string.
 *
 * When an i18n text object is provided, browser `navigator.languages` are matched against the available BCP 47 keys.
 * If no exact lookup result is found, a primary-language match is attempted before using the `fallback` field.
 *
 * @param text - Plain text or i18n text object.
 * @returns Resolved display string, or an empty string for missing text.
 */
export function resolveText(text: srk.Text | undefined): string {
  if (text === undefined) {
    return '';
  }
  if (typeof text === 'string') {
    return text;
  } else {
    const langs = Object.keys(text)
      .filter((k) => k && k !== 'fallback')
      .sort()
      .reverse();
    const userLangs = (typeof navigator !== 'undefined' && [...navigator.languages]) || [];
    const usingLang =
      langLookup(userLangs, langs) ||
      userLangs
        .map((lang) => lang.split('-')[0])
        .map((primaryLang) => langs.find((lang) => lang === primaryLang || lang.startsWith(`${primaryLang}-`)))
        .find(Boolean) ||
      '';
    return text[usingLang] ?? text.fallback ?? '';
  }
}

/**
 * Parse a contributor string into name, email, and URL parts.
 *
 * @param contributor - Contributor string in npm-style `name <email> (url)` form.
 * @returns Parsed contributor object, or `null` for missing input.
 * @example
 * 'name <mail@example.com> (http://example.com)' -> { name: 'name', email: 'mail@example.com', url: 'http://example.com' }
 * 'name' -> { name: 'name' }
 * 'name <mail@example.com>' -> { name: 'name', email: 'mail@example.com' }
 * 'name (http://example.com)' -> { name: 'name', url: 'http://example.com' }
 * 'John Smith (http://example.com)' -> { name: 'John Smith', url: 'http://example.com' }
 */
export function resolveContributor(
  contributor: srk.Contributor | undefined,
): { name: string; email?: string; url?: string } | null {
  if (!contributor) {
    return null;
  }

  let name = '';
  let email: string | undefined;
  let url: string | undefined;
  const words = contributor.split(' ').map((s) => s.trim());
  let index = words.length - 1;
  while (index > 0) {
    const word = words[index];
    if (word.startsWith('<') && word.endsWith('>')) {
      email = word.slice(1, -1);
      index--;
      continue;
    }
    if (word.startsWith('(') && word.endsWith(')')) {
      url = word.slice(1, -1);
      index--;
      continue;
    }
    break;
  }
  name = words.slice(0, index + 1).join(' ');
  return { name, email, url };
}

/**
 * Normalize an srk color value to a CSS color string.
 *
 * @param color - Color string, or an RGBA tuple supported by older data.
 * @returns CSS color string, or `undefined` when no color is provided.
 */
export function resolveColor(color: srk.Color) {
  if (Array.isArray(color)) {
    return `rgba(${color[0]},${color[1]},${color[2]},${color[3]})`;
  } else if (color) {
    return color;
  }
  return undefined;
}

/**
 * Resolve a theme color into explicit light and dark theme values.
 *
 * @param themeColor - Single color string or `{ light, dark }` theme color object.
 * @returns Theme color map keyed by `EnumTheme.light` and `EnumTheme.dark`.
 */
export function resolveThemeColor(themeColor: srk.ThemeColor): ThemeColor {
  let light = resolveColor(typeof themeColor === 'string' ? themeColor : themeColor.light);
  let dark = resolveColor(typeof themeColor === 'string' ? themeColor : themeColor.dark);
  return {
    [EnumTheme.light]: light,
    [EnumTheme.dark]: dark,
  };
}

/**
 * Resolve an srk style object into explicit light/dark text and background colors.
 *
 * If a background color is provided without a text color, a readable foreground color is selected automatically.
 *
 * @param style - srk style object.
 * @returns Resolved text and background theme colors.
 */
export function resolveStyle(style: srk.Style) {
  const { textColor, backgroundColor } = style;
  let usingTextColor: typeof textColor = textColor;
  // 未指定前景色时，尝试自动适配
  if (backgroundColor && !textColor) {
    if (typeof backgroundColor === 'string') {
      usingTextColor = TEXTColor.findTextColor(backgroundColor);
    } else {
      const { light, dark } = backgroundColor;
      usingTextColor = {
        light: light && TEXTColor.findTextColor(light),
        dark: dark && TEXTColor.findTextColor(dark),
      };
    }
  }
  const textThemeColor = resolveThemeColor(usingTextColor || '');
  const backgroundThemeColor = resolveThemeColor(backgroundColor || '');
  return {
    textColor: textThemeColor,
    backgroundColor: backgroundThemeColor,
  };
}

/**
 * Resolve the marker definitions attached to a user.
 *
 * The modern `user.markers` array takes precedence over the deprecated `user.marker` field.
 *
 * @param user - User object.
 * @param markersConfig - Available ranklist marker definitions.
 * @returns Marker objects found in `markersConfig`, preserving the user's marker order.
 */
export function resolveUserMarkers(user: srk.User, markersConfig: srk.Marker[] | undefined): srk.Marker[] {
  if (!user) {
    return [];
  }
  const userMarkers = (Array.isArray(user.markers) ? user.markers : [user.marker])
    .filter(Boolean)
    .map((marker) => (markersConfig || []).find((m) => m.id === marker))
    .filter(Boolean) as srk.Marker[];
  return userMarkers;
}

File: go/resolvers.go
package srkutils

import (
	"fmt"
	"regexp"
	"sort"
	"strings"
)

func ResolveText(text any, languages []string) string {
	if text == nil {
		return ""
	}
	if value, ok := text.(string); ok {
		return value
	}
	values := asMap(text)
	langs := []string{}
	for key := range values {
		if key != "" && key != "fallback" {
			langs = append(langs, key)
		}
	}
	sort.Sort(sort.Reverse(sort.StringSlice(langs)))
	usingLang := ""
	for _, lang := range languages {
		for _, candidate := range langs {
			if candidate == lang {
				usingLang = candidate
				break
			}
		}
		if usingLang != "" {
			break
		}
	}
	if usingLang == "" {
		for _, lang := range languages {
			primary := strings.Split(lang, "-")[0]
			for _, candidate := range langs {
				if candidate == primary || strings.HasPrefix(candidate, primary+"-") {
					usingLang = candidate
					break
				}
			}
			if usingLang != "" {
				break
			}
		}
	}
	if value, ok := values[usingLang].(string); ok {
		return value
	}
	if value, ok := values["fallback"].(string); ok {
		return value
	}
	return ""
}

func ResolveContributor(contributor string) map[string]any {
	if contributor == "" {
		return nil
	}
	words := strings.Fields(contributor)
	index := len(words) - 1
	email := ""
	url := ""
	for index > 0 {
		word := words[index]
		if strings.HasPrefix(word, "<") && strings.HasSuffix(word, ">") {
			email = word[1 : len(word)-1]
			index--
			continue
		}
		if strings.HasPrefix(word, "(") && strings.HasSuffix(word, ")") {
			url = word[1 : len(word)-1]
			index--
			continue
		}
		break
	}
	result := map[string]any{"name": strings.Join(words[:index+1], " ")}
	if email != "" {
		result["email"] = email
	}
	if url != "" {
		result["url"] = url
	}
	return result
}

func ResolveColor(color any) any {
	if raw := asSlice(color); len(raw) >= 4 {
		return fmt.Sprintf("rgba(%v,%v,%v,%v)", raw[0], raw[1], raw[2], raw[3])
	}
	if value, ok := color.(string); ok {
		if value != "" {
			return value
		}
	}
	return nil
}

func ResolveThemeColor(themeColor any) map[string]any {
	if value, ok := themeColor.(string); ok {
		color := ResolveColor(value)
		return map[string]any{"light": color, "dark": color}
	}
	values := asMap(themeColor)
	return map[string]any{"light": ResolveColor(values["light"]), "dark": ResolveColor(values["dark"])}
}

func parseColorRGB(color string) (float64, float64, float64) {
	if strings.HasPrefix(color, "#") && len(color) == 4 {
		r, _ := strconvParseHex(strings.Repeat(color[1:2], 2))
		g, _ := strconvParseHex(strings.Repeat(color[2:3], 2))
		b, _ := strconvParseHex(strings.Repeat(color[3:4], 2))
		return r, g, b
	}
	if strings.HasPrefix(color, "#") && len(color) == 7 {
		r, _ := strconvParseHex(color[1:3])
		g, _ := strconvParseHex(color[3:5])
		b, _ := strconvParseHex(color[5:7])
		return r, g, b
	}
	re := regexp.MustCompile(`rgba?\(([^,]+),([^,]+),([^,\)]+)`)
	match := re.FindStringSubmatch(color)
	if len(match) == 4 {
		return numberValue(match[1]), numberValue(match[2]), numberValue(match[3])
	}
	return 255, 255, 255
}

func strconvParseHex(value string) (float64, error) {
	var parsed int64
	_, err := fmt.Sscanf(value, "%x", &parsed)
	return float64(parsed), err
}

func autoTextColor(backgroundColor string) string {
	red, green, blue := parseColorRGB(backgroundColor)
	if 0.213*red+0.715*green+0.072*blue > 255/2 {
		return "#000000"
	}
	return "#ffffff"
}

func ResolveStyle(style map[string]any) map[string]any {
	textColor := style["textColor"]
	backgroundColor := style["backgroundColor"]
	usingTextColor := textColor
	if backgroundColor != nil && textColor == nil {
		if value, ok := backgroundColor.(string); ok {
			usingTextColor = autoTextColor(value)
		} else {
			theme := asMap(backgroundColor)
			result := map[string]any{}
			if theme["light"] != nil {
				result["light"] = autoTextColor(stringValue(theme["light"]))
			}
			if theme["dark"] != nil {
				result["dark"] = autoTextColor(stringValue(theme["dark"]))
			}
			usingTextColor = result
		}
	}
	return map[string]any{
		"textColor":       ResolveThemeColor(firstNonNil(usingTextColor, "")),
		"backgroundColor": ResolveThemeColor(firstNonNil(backgroundColor, "")),
	}
}

func firstNonNil(value any, fallback any) any {
	if value == nil {
		return fallback
	}
	return value
}

func ResolveUserMarkers(user map[string]any, markersConfig []map[string]any) []map[string]any {
	if user == nil {
		return []map[string]any{}
	}
	var userMarkers []any
	if markers, ok := user["markers"]; ok {
		userMarkers = asSlice(markers)
	} else {
		userMarkers = []any{user["marker"]}
	}
	result := []map[string]any{}
	for _, markerID := range userMarkers {
		if markerID == nil {
			continue
		}
		for _, marker := range markersConfig {
			if marker["id"] == markerID {
				result = append(result, marker)
				break
			}
		}
	}
	return result
}
